### DISCUSSION TITLE
Reuse types.union and TS union

### DISCUSSION THREAD

**Discussion by sandbox4013** [mobxjs/mobx-state-tree#1752] [Q&A]
```
type SomeStatus = 'active' | 'suspended' | 'pending'
// type SomeStatus = typeof Some['status']

types.model('Some', {
    status: types.union(
        types.literal('active'),
        types.literal('suspended'),
        types.literal('pending')
    )
    // status: SomeStatus
})

```

I want to have one 'source of truth' - either model field, or union type itself
How can I do it?

**Comment by BATCOH** [comment]
For things like statuses, you can use the `types.enumeration`

> [`types.enumeration(name?, options: string[])`](https://mobx-state-tree.js.org/API/#enumeration) creates an enumeration. This method is a shorthand for a union of string literals. If you are using typescript and want to create a type based on an string enum (e.g. `enum Color { ... }`) then use `types.enumeration<Color>("Color", Object.values(Color))`, where the "Color" name argument is optional.

**Reply by sandbox4013** [reply]
> For things like statuses, you can use the `types.enumeration`
> 
> > [`types.enumeration(name?, options: string[])`](https://mobx-state-tree.js.org/API/#enumeration) creates an enumeration. This method is a shorthand for a union of string literals. If you are using typescript and want to create a type based on an string enum (e.g. `enum Color { ... }`) then use `types.enumeration<Color>("Color", Object.values(Color))`, where the "Color" name argument is optional.

I want to create a type base on an string union type!

**Reply by elektronik2k5** [reply]
You can't cause the way MST works is in the opposite direction: you declare types in MST primitives - from which TS types are inferred. 
Even if we somehow (in theory) manage to support the opposite, it won't work for JS based codebases, which are otherwise fully supported by MST - by design.

**Comment by YuanRuQian** [comment]
How about a self-implemented `enumeration` to support readonly array? 
So you can always use the `SOME_STATUS` as the single source of truth.

```typescript
const SOME_STATUS = ['active', 'suspended', 'pending'] as const

type SomeStatus = typeof SOME_STATUS[number]
// type SomeStatus = "active" | "suspended" | "pending"

function enumeration<T extends string>(
  name: string,
  values: T[] | readonly T[]
): ISimpleType<T> {
  const source = Array.isArray(values) ? values : Object.keys(values)
  const literals = source.map(value => types.literal(value))
  const union = types.union(...literals)
  union.name = name
  return union
}

const SomeStatus = types.model('SomeStatus', {
  status: enumeration('some', SOME_STATUS),
})

const someStatus = SomeStatus.create({
  status: 'active',
  // (property) status: "active" | "suspended" | "pending"
})

```

PS This is just my personal workaround demo, I haven't thought about the error handling or other more serious things under the hood.

**Reply by sandbox4013** [reply]
How 'number' in line
`type SomeStatus = typeof SOME_STATUS[number]`
works?

**Reply by YuanRuQian** [reply]
It's the `as const` const-assertion really works the magic here.

I think this [StackOverflow answer](https://stackoverflow.com/a/66993654/10340117) explains this pretty well and even much better than the official docs on this subject.

I hope the link above helps!

**Reply by sandbox4013** [reply]
thanks!
